tensorflow/contrib/learn/python/learn/estimators/estimator.py

```diff
@@ -710,28 +710,29 @@ def _get_feature_ops_from_example(self, examples_batch):
     return tensor_signature.create_example_parser_from_signatures(
         self._features_info, examples_batch)
 
-  def _check_inputs(self, features, labels):
-    if self._features_info is not None:
-      logging.debug('Given features: %s, required signatures: %s.',
-                    str(features), str(self._features_info))
+  def _check_inputs(self, features, targets, mode):
+    if mode in self._features_info:
+      logging.debug('Given features for %s: %s, required signatures: %s.',
+                    mode, str(features), str(self._features_info[mode]))
+
       if not tensor_signature.tensors_compatible(features, self._features_info):
-        raise ValueError('Features are incompatible with given information. '
+        raise ValueError('Features are incompatible for %s with given information. '
                          'Given features: %s, required signatures: %s.' %
-                         (str(features), str(self._features_info)))
+                         (mode, str(features), str(self._features_info[mode])))
```

Comment:
@martinwicke - I am bit confused about this comment. train/eval/predict call `_check_inputs` inside their function call. If the features don't already have the signature, they are created. I don't see how "it is not allowed to pass different features to train/eval/predict". `self._features_info[mode]` and `self._labels_info[mode]` binds the signature for particular mode (namely 'train', 'eval' and 'predict'). One is free to use inputs with same signature across these modes though. May be I completely missed your point, please elaborate.